packages/block-editor/src/components/url-input/style.scss

```diff
@@ -94,7 +94,7 @@ $input-size: 300px;
 
 	&:focus,
 	&.is-selected {
-		background: color(theme(primary) shade(15%));
+		background: theme(primary, shade-20);
```

Comment:
@ryelle Making sure I understand what this is doing - is the `theme` function here equivalent to calling `var( --color-primary--shade-20 )`?

Reply:
If yes, then with the "additional level of abstraction" I've mentioned, this would be `var( --block-editor-url-input__suggestion-focus-background )` and a the `--color-primary--shade-20`  assigned in `body.theme-name`.

Reply:
Yeah, `theme` generates the `var(…)` code. Just to clarify, with the abstraction layer, would each theme be able to control both tokens & assignments, like this?

```css
body.theme-red {
    --color-primary: red;
    --color-primary--shade-20: darkred;

    --block-editor-url-input__suggestion-focus-background: var( --color-primary );
}

body.theme-blue {
    --color-primary: blue;
    --color-primary--shade-20: darkblue;

    --block-editor-url-input__suggestion-focus-background: var( --color-primary--shade-20 )
}
```

Reply:
My first thought is no, only tokens - controlling the assignments would be potentially too flexible and it could become a mess dealing with specific logic for color contrast. I'm not sure on that though!

Reply:
Maybe for specific themes, as needed - but then again, a principle in using the tokens approach is to have restrictions and decisions like that already made. I feel like an idea workflow would be:

1. Set base token values and tweak tint and shade assignments (we might want those to have names like `lightest` to `darkest` instead of numbers
2. Pass some sort of test that checks color contrast ratios in the theme

Reply:
So it would be something like:

```
{
    themes: {
		default: {
			primary: '#f00',
			secondary: '#f00',
			accent: '#f00',
			error: '#f00',
			success: '#f00',
			tints: {
				lightest: '95',
				lighter: '90',
				light: '72',
				dark: '40',
				darker: '20',
				darkest: '5'
			},
			shades: {
				shallowest: '#',
				shallower: '#',
				shallow: '#',
				deep: '#',
				deeper: '#',
				deepest: '#'
			}
		},
		blue: {
			primary: '#f00',
			secondary: '#f00',
			accent: '#f00',
			error: '#f00',
			success: '#f00',
			tints: {
				lightest: '99',
				lighter: '82',
				light: '72',
				dark: '30',
				darker: '21',
				darkest: '8'
			},
			shades: {
				shallowest: '#',
				shallower: '#',
				shallow: '#',
				deep: '#',
				deeper: '#',
				deepest: '#'
			}
		}
	},
}
```
{
    defaults: { primary: '#00f' },
    themes: { 
red: { primary: '#f00' } },
}
```

(where the `'#'` is a descending value for shades, varying for each theme.

Reply:
Actually...this is potentially a different idea than what I initially mentioned with the UI-named abstraction - the PostCSS plugin would generate those custom properties for us, then we apply them like this in the stylesheet:

```

body.theme-blue {
	/* all the custom properties generated based on above spec for blue */
}

/* Not sure what this is exactly, but we to attach the colors to the UI-named properties */
.some-other-context {
	--button-background-color: var( --color-primary--shade-darkest );
	--button-text-color: var( --color-primary--shade-lightest );
	--button-focus-background-color: var( --color-secondary--shade-darkest );
	--button-focus-text-color: var( --color-secondary--shade-lightest );
}

.components-button {
	background-color: var( --button-background-color );
	color: var( --button-text-color );

	// state bindings for the colors -
	// Fallbacks are perhaps in a separate stylesheet / only in the build
	&:focus {
		--button-background-color: var( --button-focus-background-color );
		--button-text-color: var( --button-focus-text-color );
	}
}
```